### PR TITLE
[SPARK-33677][SQL] Skip LikeSimplification rule if pattern contains any escapeChar

### PR DESCRIPTION
**Pull request by luluorta**
### What changes were proposed in this pull request?
`LikeSimplification` rule does not work correctly for many cases that have patterns containing escape characters, for example:

`SELECT s LIKE 'm%aca' ESCAPE '%' FROM t`
`SELECT s LIKE 'maacaa' ESCAPE 'a' FROM t`

For simpilicy, this PR makes this rule just be skipped if `pattern` contains any `escapeChar`.

### Why are the changes needed?
Result corrupt.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added Unit test.

### CODE REVIEW

**Comment by luluorta** [comment]
cc @cloud-fan @leanken

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -543,13 +543,14 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
   def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
-    case Like(input, Literal(pattern, StringType), escapeChar) =>
+    case l @ Like(input, Literal(pattern, StringType), escapeChar) =>
       if (pattern == null) {
         // If pattern is null, return null value directly, since "col like null" == null.
         Literal(null, BooleanType)
       } else {
         val escapeStr = String.valueOf(escapeChar)
         pattern.toString match {
+          case _ if escapeChar == '_' || escapeChar == '%' => l
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:553]: Please add tests in `LikeSimplificationSuite`, too? Btw, does this simplification work correctly if `escapeChar` is not a default one other than `_` and `%`?

Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:553]: You are right! I think this rule works correctly only if `pattern` is valid and `escapeChar` does not escape itself. I will fix this and change the title/description.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:553]: oh... I see...

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -3718,6 +3718,14 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       }
     }
   }
+
+  test("SPARK-33677: LikeSimplification should be skipped if escape is a wildcard character") {
+    val e = intercept[AnalysisException] {
+      sql("SELECT string(rand()) LIKE 'm%aca' ESCAPE '%'").collect()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: so the problem is, we should only apply like simplification if the pattern is valid?

Comment (luluorta) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: This rule may not work even for the case that has a valid pattern: `SELECT s LIKE 'maacaa' ESCAPE 'a' FROM t`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -543,27 +543,27 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
-    case Like(input, Literal(pattern, StringType), escapeChar) =>
+    case l @ Like(input, Literal(pattern, StringType), escapeChar) =>
       if (pattern == null) {
         // If pattern is null, return null value directly, since "col like null" == null.
         Literal(null, BooleanType)
       } else {
-        val escapeStr = String.valueOf(escapeChar)
         pattern.toString match {
-          case startsWith(prefix) if !prefix.endsWith(escapeStr) =>
+          case p if p.contains(escapeChar) => l
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:552]: correct me if I was wrong: we need to make sure this rule doesn't change result. If the pattern is invalid, `Like` should fail. However, it's expensive to validate the pattern (need to compile it). Here we use `p.contains(escapeChar)` as a shortcut to know if the pattern might be invalid.

Can we add some comments to explain it?

Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:558]: done

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -3718,6 +3718,20 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       }
     }
   }
+
+  test("SPARK-33677: LikeSimplification should be skipped if pattern contains any escapeChar") {
+    withTable("string_tbl") {
+      sql("CREATE TABLE string_tbl USING parquet SELECT 'm@ca' AS s")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: nit: could you use a temporary table for better test performance?

**Review by maropu — APPROVED** [review]
Nice catch!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: there is no temporary table in spark, you mean temporary view?

Comment (luluorta) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: Unfortunately `Like` will be constant folded and not passed to `LikeSimplification` If using temp view here.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: Oh, yea, I meant a temporary view...

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3724]: @luluorta have you tried temp view? AFAIK `ConvertToLocalRelation` is disabled in tests.

Comment (luluorta) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: It works if I use DataFrame to create temp view.

**Comment by cloud-fan** [comment]
retest this please

**Comment by maropu** [comment]
Thanks! Merged to master/branch-3.1/branch-3.0.

**Pull request closed without merging**